src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
+        } else {
+            oldHead.prev = newNode;
+        }
+        size++;
+    }
+
+    private T removeHead() {
+        final T oldValue = head.value;
+        if (size > 1) {
+            head = head.next;
+            head.prev = null;
+            size--;
+        } else {
+            head.next = null;
+            head.value = null;
+            size = 0;
+        }
+        return oldValue;
+    }
+
+    private T removeTail() {
+        final T oldValue = tail.value;
+        tail = tail.prev;
+        tail.next = null;
+        size--;
+        return oldValue;
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+            size++;
+            return;
+        }
+        Node<T> last = tail;
+        Node<T> newNode = new Node<>(last, value, null);
+        tail = newNode;
+        if (last == null) {
+            head = newNode;
+        } else {
+            last.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (!(index >= 0 && index <= size)) {
+            throw new IndexOutOfBoundsException(
+                    "Index passed to the method is negative or "
+                            + "bigger than size of MyLinkedList. "
+                            + "Index: " + index + ", Size: " + size);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            setHead(value);
+            return;
+        }
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+
+        Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+        currentNode.prev.next = newNode;
+        currentNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(
+                    "Index passed to the method is negative or "
+                            + "bigger than size of MyLinkedList. "
+                            + "Index: " + index + ", Size: " + size);
+        }
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(
+                    "Index passed to the method is negative or "
+                            + "bigger than size of MyLinkedList. "
+                            + "Index: " + index + ", Size: " + size);
+        }
```

Comment:
you have the same check in several places, create a separate method and move this logic there